### PR TITLE
Use react-loadable to render routes

### LINKED ISSUE #152: Use react-loadable to render routes

### PR DESCRIPTION
**Pull request by fiennyangeln**
Closes #152

### CODE REVIEW

File: src/App.jsx [hunk 1]
```diff
@@ -31,6 +42,9 @@ persistCache({
   },
 })
 export default class App extends Component {
+  state = {
+    error: null,
+  };
```
Comment (helfi92) [inline src/App.jsx]: Add newline.

File: src/App.jsx [hunk 2]
```diff
@@ -10,12 +10,23 @@ import { persistCache } from 'apollo-cache-persist';
...
+  {
+    component: loadable(() => import('./views/Projects')),
+    path: '/',
+    exact: true,
+  },
+  {
+    component: loadable(() => import('./views/Project')),
+    path: '/projects/:project',
+  },
+];
```
Comment (helfi92) [inline src/App.jsx]: Put this inside `routes.jsx` similar to how tc-web does it.

File: src/utils/loadable.jsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    if (error) {
+      throw error;
+    } else if (timedOut || pastDelay) {
+      return (
+        <CircularProgress
+          size={50}
+          classes={{
+            circleIndeterminate: classes.spinner,
+          }}
+        />
```
Comment (helfi92) [inline src/utils/loadable.jsx]: Use the spinner from `components/Spinner`.

File: src/utils/loadable.jsx [hunk 4]
```diff
@@ -15,25 +15,15 @@ import CircularProgress from '@material-ui/core/CircularProgress';
...
     } else if (timedOut || pastDelay) {
-      return (
-        <CircularProgress
-          size={50}
-          classes={{
-            circleIndeterminate: classes.spinner,
-          }}
-        />
-      );
+      return <Spinner size={50} />;
```
Comment (helfi92) [inline src/utils/loadable.jsx]: Remove `size` so we make sure the spinner looks identical to the one we use in the project view.

File: src/components/Spinner/index.jsx [hunk 5]
```diff
@@ -10,11 +10,11 @@ import classNames from 'classnames';
 })
 export default class Spinner extends PureComponent {
   render() {
-    const { classes, className } = this.props;
+    const { classes, className, size } = this.props;
```
Comment (helfi92) [inline src/components/Spinner/index.jsx]: Remove `size`.

File: src/utils/loadable.jsx [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+  },
+}))
+class Loading extends PureComponent {
+  content() {
+    const { error, timedOut, pastDelay } = this.props;
+
+    if (error) {
+      throw error;
+    } else if (timedOut || pastDelay) {
+      return <Spinner />;
```
Comment (helfi92) [inline src/utils/loadable.jsx]: The spinner doesn't seem to be centered. Could you center it, please?

File: src/utils/loadable.jsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+      throw error;
+    } else if (timedOut || pastDelay) {
+      return <Spinner className={classes.view} />;
+    }
+
+    return null;
+  }
+
+  render() {
+    return <Fragment>{this.content()}</Fragment>;
```
Comment (helfi92) [inline src/utils/loadable.jsx]: This should be a `div`.

File: src/utils/loadable.jsx [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+  },
+}))
+class Loading extends PureComponent {
+  content() {
+    const { classes, error, timedOut, pastDelay } = this.props;
+
+    if (error) {
+      throw error;
+    } else if (timedOut || pastDelay) {
+      return <Spinner className={classes.view} />;
```
Comment (helfi92) [inline src/utils/loadable.jsx]: I would have the parent be a `div` and let it position the spinner. In other words, the render method should be `<div className={classes.view}>{this.content()}</div>`.

File: src/utils/loadable.jsx [hunk 9]
```diff
@@ -7,37 +7,32 @@ import Spinner from '../components/Spinner';
   view: {
     textAlign: 'center',
     margin: 100,
-    width: '100%',
```
Comment (helfi92) [inline src/utils/loadable.jsx]: Assuming the class is on the `div`, `width` of 100% should still be there.

File: src/utils/loadable.jsx [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    marginTop: 60,
+  },
+}))
+class Loading extends PureComponent {
+  content() {
+    const { error, timedOut, pastDelay } = this.props;
+
+    if (error) {
+      throw error;
+    } else if (timedOut || pastDelay || !error) {
```
Comment (helfi92) [inline src/utils/loadable.jsx]: I don't think we need `|| !error` since the if statement just above checks for it.

**Comment by fiennyangeln** [comment]
Ok! 😄

**Pull request merged**